Reduce instance supplier use to appease AOT

onobc opened the pull request:
* Remove the use of instance suppliers on bean definitions that are processed during the AOT phase.

* The remaining areas that use instance suppliers do so at runtime and do not use reflection, but instead
are passed the configured bean to register.

See https://github.com/spring-cloud/spring-cloud-stream/issues/2655

onobc commented on spring-integration-core/src/main/java/org/springframework/integration/config/DefaultConfiguringBeanFactoryPostProcessor.java:
```diff
@@ -173,8 +174,7 @@ private void registerNullChannel() {
 			do {
-				if (beanFactoryToUse instanceof ConfigurableListableBeanFactory) {
-					ConfigurableListableBeanFactory listable = (ConfigurableListableBeanFactory) beanFactoryToUse;
+				if (beanFactoryToUse instanceof ConfigurableListableBeanFactory listable) {
```
**[UNRELATED]** I agreed w/ IDEA suggestion

onobc commented on spring-integration-core/src/main/java/org/springframework/integration/dsl/context/IntegrationFlowBeanPostProcessor.java:
```diff
@@ -421,8 +420,7 @@ private boolean noBeanPresentForComponent(Object instance, String parentBeanName
 		return !this.beanFactory.getBeansOfType(instance.getClass(), false, false)
-				.values()
-				.contains(instance);
+				.containsValue(instance);
```
**[UNRELATED]** I like this new API

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/AbstractMethodAnnotationPostProcessor.java:
```diff
@@ -577,7 +577,7 @@ protected AbstractEndpoint createEndpoint(MessageHandler handler, @SuppressWarni
 				if (e.getCause() instanceof NoSuchBeanDefinitionException) {
 					this.definitionRegistry.registerBeanDefinition(inputChannelName,
-							new RootBeanDefinition(DirectChannel.class, DirectChannel::new));
+							new RootBeanDefinition(DirectChannel.class));
```
Revert this one, please: it is used at runtime and not in AOT at all.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/InboundChannelAdapterAnnotationPostProcessor.java:
```diff
@@ -122,12 +121,12 @@ public Object postProcess(Object bean, String beanName, Method method, List<Anno
 		String messageSourceBeanName = generateHandlerBeanName(beanName, method);
-		getDefinitionRegistry().registerBeanDefinition(messageSourceBeanName,
-				new RootBeanDefinition(MethodInvokingMessageSource.class, () -> methodInvokingMessageSource));
+		getDefinitionRegistry().registerBeanDefinition(messageSourceBeanName, beanDefinition);
```
Same here: not AOT, but rather runtime stuff.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/dsl/context/IntegrationFlowBeanPostProcessor.java:
```diff
@@ -408,8 +408,7 @@ private boolean noBeanPresentForComponent(Object instance, String parentBeanName
 					AbstractBeanDefinition beanDefinition =
-							BeanDefinitionBuilder.genericBeanDefinition((Class<Object>) instance.getClass(),
-											() -> instance)
+							BeanDefinitionBuilder.genericBeanDefinition(instance.getClass())
```
Revert this one. You essentially making a bug here: a bean definition is going to be based on a new instance of the class, but not provided one as intention.
More over this class does its stuff at runtime, so the change is redundant.

artembilan reviewed (changes requested):
Thank you, Chris!
Just a couple requests to restore existing behavior in your change.
The mostly important concern that we would like to deal with existing instances in those bean definitions, but not creating new as it would be with a default class-based configuration.

onobc committed:
Reduce instance supplier use to appease AOT

* Remove the use of instance suppliers on bean definitions
that are processed during the AOT phase.

* The remaining areas that use instance suppliers do so
at runtime and do not use reflection, but instead
are passed the configured bean to register.

See https://github.com/spring-cloud/spring-cloud-stream/issues/2655

artembilan commented:
... and cherry-picked to `6.0.x`
